File: php/tests/keylog.test.php
<?php
// Key-log emitter tests (security.md §8), component-level: PHP cannot run two
// nodes in one process, so we run a full in-process handshake to get both ends'
// sessions, emit each end's key-log via the node's writeKeylog, and assert the
// two logs agree on the directional keys and transcript-hash label — proof the
// emitted keys are the real shared session keys and the role→direction mapping
// is correct. The cross-node key-log + bonemesh-inspect round-trip is tier 10.
use Bonemesh\Handshake;
use Bonemesh\Frame;
use Bonemesh\Cert;
use Bonemesh\Crypto;
use Bonemesh\Node;

const KL_MESH = 'acme-prod';

function kl_issue(array $root, string $label): array
{
    $kp = Crypto::mldsa65Generate();
    $cert = ca_sign_cert($root, Cert::build(KL_MESH, $label, $kp['pub'], 1000, 2000));
    return ['cert' => $cert, 'idPrivate' => $kp['priv']];
}

function kl_dec(string $frame): array
{
    return Frame::classify($frame, Frame::HANDSHAKE_CAP)['obj'];
}

// A full three-message handshake; returns [initiatorSession, responderSession].
function kl_handshake(): array
{
    $root = ca_root();
    $i = kl_issue($root, 'initiator');
    $r = kl_issue($root, 'responder');
    $init = Handshake::initiator(KL_MESH, $root['pubRaw'], 1500, $i['cert'], $i['idPrivate']);
    $resp = Handshake::responder(KL_MESH, $root['pubRaw'], 1500, $r['cert'], $r['idPrivate']);
    $m2 = kl_dec($resp->readMessage1WriteMessage2(kl_dec($init->writeMessage1())));
    $m3 = kl_dec($init->readMessage2WriteMessage3($m2));
    $resp->readMessage3($m3);
    return [$init->session(), $resp->session()];
}

// Invoke the node's private writeKeylog with an explicit key-log path.
function kl_write(string $path, bool $initiator, array $sess): void
{
    $node = new Node(['label' => 'x', 'mesh' => 'm', 'rootPublic' => '', 'cert' => [], 'idPrivate' => '']);
    $ref = new ReflectionObject($node);
    $tunProp = $ref->getProperty('tun');
    $tunProp->setAccessible(true);
    $tun = $tunProp->getValue($node);
    $tun['keylogPath'] = $path;
    $tunProp->setValue($node, $tun);
    $m = $ref->getMethod('writeKeylog');
    $m->setAccessible(true);
    $m->invoke($node, 0, $initiator, $sess);
}

function kl_parse(string $path): array
{
    $out = [];
    foreach (explode("\n", trim(file_get_contents($path))) as $ln) {
        if ($ln === '' || $ln[0] === '#') {
            continue;
        }
        $f = explode(' ', $ln);
        if (count($f) !== 3) {
            continue;
        }
        $dir = explode('_', $f[0])[1]; // BMX3_<DIR>_TRAFFIC_0 -> DIR
        $out[$dir] = ['th' => $f[1], 'key' => $f[2]];
    }
    return $out;
}

test('keylog emits agreeing directional keys and transcript hash for both ends', function () {
    [$is, $rs] = kl_handshake();
    $fa = tempnam(sys_get_temp_dir(), 'kla');
    $fb = tempnam(sys_get_temp_dir(), 'klb');
    kl_write($fa, true, $is);   // initiator's log
    kl_write($fb, false, $rs);  // responder's log

    $a = kl_parse($fa);
    $b = kl_parse($fb);
    @unlink($fa);
    @unlink($fb);

    assertNotNull($a['I2R'] ?? null, 'initiator wrote an I2R line');
    assertNotNull($b['I2R'] ?? null, 'responder wrote an I2R line');
    assertEq(64, strlen($a['I2R']['key']), 'key is 32 bytes of hex');
    // Both ends must name the same key for each absolute direction — this fails
    // if the responder's role→direction swap is wrong.
    assertEq($a['I2R']['key'], $b['I2R']['key'], 'I2R key disagrees between ends');
    assertEq($a['R2I']['key'], $b['R2I']['key'], 'R2I key disagrees between ends');
    assertEq($a['I2R']['th'], $b['I2R']['th'], 'transcript-hash label disagrees');
    assertTrue($a['I2R']['key'] !== $a['R2I']['key'], 'the two directions must use different keys');
});

test('keylog is silent when BONEMESH_KEYLOG is unset', function () {
    [$is, ] = kl_handshake();
    $node = new Node(['label' => 'x', 'mesh' => 'm', 'rootPublic' => '', 'cert' => [], 'idPrivate' => '']);
    $ref = new ReflectionObject($node);
    // Default tun has keylogPath '' (env unset in tests) — writeKeylog must not throw or write.
    $m = $ref->getMethod('writeKeylog');
    $m->setAccessible(true);
    $m->invoke($node, 0, true, $is); // no path configured -> no-op, no error
    assertTrue(true);
});

File: rust/tests/keylog_test.rs
// BONEMESH_KEYLOG emitter over real loopback nodes (security.md §8). Its own
// test binary so the process-wide env var cannot bleed into other tests. Both
// ends write their directional transport keys; because the role→direction
// mapping is correct, the two ends agree on the I2R key, the R2I key, and the
// transcript-hash label — proof the emitted keys are the real shared session
// keys.
use std::time::{Duration, Instant};

use bonemesh::node::{Config, Node};
use bonemesh::{cert, crypto};

const MESH: &str = "acme-prod";
const NOW: i64 = 1_788_600_000;

fn config(root_priv: &[u8; 32], root_pub: &[u8], label: &str) -> Config {
    let (pub_key, priv_seed) = crypto::mldsa65_generate();
    let c = cert::sign(cert::build(MESH, label, &pub_key, NOW - 100, NOW + 100_000), root_priv);
    Config {
        label: label.to_string(),
        mesh: MESH.to_string(),
        root_public: root_pub.to_vec(),
        cert: c,
        id_private: priv_seed,
    }
}

// dir -> set of (th, key) seen at epoch 0.
fn parse(path: &str) -> std::collections::HashMap<String, Vec<(String, String)>> {
    let mut out: std::collections::HashMap<String, Vec<(String, String)>> = Default::default();
    let raw = std::fs::read_to_string(path).unwrap_or_default();
    for ln in raw.lines() {
        let f: Vec<&str> = ln.split_whitespace().collect();
        if f.len() != 3 || !f[0].ends_with("_TRAFFIC_0") {
            continue;
        }
        let dir = f[0].trim_start_matches("BMX3_").trim_end_matches("_TRAFFIC_0").to_string();
        out.entry(dir).or_default().push((f[1].to_string(), f[2].to_string()));
    }
    out
}

#[test]
fn keylog_emits_agreeing_directional_keys() {
    let path = std::env::temp_dir()
        .join("bonemesh_keylog_test.klog")
        .to_string_lossy()
        .to_string();
    let _ = std::fs::remove_file(&path);
    std::env::set_var("BONEMESH_KEYLOG", &path);

    let (root_pub, root_priv) = crypto::mldsa87_generate();
    let alpha = Node::start(config(&root_priv, &root_pub, "alpha"), 0).unwrap();
    let beta = Node::start(config(&root_priv, &root_pub, "beta"), 0).unwrap();
    alpha.connect("127.0.0.1", beta.port()).unwrap();

    // Wait for both ends to write their epoch-0 entries (4 lines total).
    let deadline = Instant::now() + Duration::from_secs(5);
    while Instant::now() < deadline {
        let n = std::fs::read_to_string(&path).unwrap_or_default().lines().count();
        if n >= 4 {
            break;
        }
        std::thread::sleep(Duration::from_millis(50));
    }

    let by_dir = parse(&path);
    for dir in ["I2R", "R2I"] {
        let entries = by_dir.get(dir).unwrap_or_else(|| panic!("no {dir} entries"));
        assert!(entries.len() >= 2, "{dir}: both ends should have written it, got {}", entries.len());
        let key0 = &entries[0].1;
        assert_eq!(key0.len(), 64, "{dir} key is not a 32-byte hex value: {key0}");
        for e in entries {
            assert_eq!(&e.1, key0, "{dir} key disagrees between ends (role→direction mapping wrong)");
            assert_eq!(e.0, entries[0].0, "{dir} transcript-hash disagrees between ends");
        }
    }
    assert_ne!(by_dir["I2R"][0].1, by_dir["R2I"][0].1, "I2R and R2I keys are identical");

    std::env::remove_var("BONEMESH_KEYLOG");
    let _ = std::fs::remove_file(&path);
    alpha.kill();
    beta.kill();
}
